File: sdk/node/src/resources/council.ts
import type { HttpClient } from '../client.js';
import type { CouncilSession, CouncilMessage } from '../types.js';

export class CouncilResource {
  constructor(private readonly http: HttpClient) {}

  run(opts: { task: string; context?: Record<string, unknown> }): Promise<{ session_id: string; status: string; task: string; created_at: string }> {
    return this.http.post('/v1/council/run', opts);
  }

  list(opts?: { limit?: number }): Promise<{ session_id: string; task: string; status: string; created_at: string }[]> {
    const qs = opts?.limit ? `?limit=${opts.limit}` : '';
    return this.http.get(`/v1/council${qs}`);
  }

  get(sessionId: string): Promise<CouncilSession> {
    return this.http.get(`/v1/council/${sessionId}`);
  }

  messages(sessionId: string): Promise<{ session_id: string; messages: CouncilMessage[] }> {
    return this.http.get(`/v1/council/${sessionId}/messages`);
  }

  /** Poll until the session reaches a terminal state. */
  async wait(
    sessionId: string,
    opts?: { pollMs?: number; timeoutMs?: number },
  ): Promise<CouncilSession> {
    const pollMs = opts?.pollMs ?? 3000;
    const timeoutMs = opts?.timeoutMs ?? 300_000; // 5 min default
    const deadline = Date.now() + timeoutMs;

    while (Date.now() < deadline) {
      const session = await this.get(sessionId);
      if (session.status === 'completed' || session.status === 'failed') {
        return session;
      }
      await sleep(pollMs);
    }
    throw new Error(`Council session ${sessionId} did not complete within ${timeoutMs}ms`);
  }

  /** run() + wait() in one call */
  async runAndWait(
    opts: { task: string; context?: Record<string, unknown> },
    waitOpts?: { pollMs?: number; timeoutMs?: number },
  ): Promise<CouncilSession> {
    const { session_id } = await this.run(opts);
    return this.wait(session_id, waitOpts);
  }
}

function sleep(ms: number) { return new Promise((r) => setTimeout(r, ms)); }

File: sdk/python/perpetual/resources/council.py
from __future__ import annotations
import time
from typing import Any
from ..http import HttpClient
from ..types import CouncilSession, PerpetualError


class CouncilResource:
    def __init__(self, http: HttpClient) -> None:
        self._http = http

    def run(self, task: str, context: dict[str, Any] | None = None) -> dict:
        body: dict[str, Any] = {"task": task}
        if context:
            body["context"] = context
        return self._http.post("/v1/council/run", body)

    def list(self, limit: int = 20) -> list[dict]:
        return self._http.get(f"/v1/council?limit={limit}")

    def get(self, session_id: str) -> CouncilSession:
        data = self._http.get(f"/v1/council/{session_id}")
        return CouncilSession(**data)

    def messages(self, session_id: str) -> dict:
        return self._http.get(f"/v1/council/{session_id}/messages")

    def wait(
        self,
        session_id: str,
        poll_secs: float = 3.0,
        timeout_secs: float = 300.0,
    ) -> CouncilSession:
        deadline = time.monotonic() + timeout_secs
        while time.monotonic() < deadline:
            session = self.get(session_id)
            if session.status in ("completed", "failed"):
                return session
            time.sleep(poll_secs)
        raise PerpetualError(
            f"Council session {session_id} did not complete within {timeout_secs}s", 0
        )

    def run_and_wait(
        self,
        task: str,
        context: dict[str, Any] | None = None,
        poll_secs: float = 3.0,
        timeout_secs: float = 300.0,
    ) -> CouncilSession:
        result = self.run(task, context)
        return self.wait(result["session_id"], poll_secs=poll_secs, timeout_secs=timeout_secs)
